scripts/base/protocols/ldap/main.zeek

```diff
@@ -0,0 +1,513 @@
+                                                     LDAP::ProtocolOpcode_SEARCH_RESULT_REFERENCE };
+
+#############################################################################
+redef record connection += {
+  ldap_proto: string &optional;
+  ldap_messages: table[int] of Message &optional;
+  ldap_searches: table[int] of Search &optional;
+};
+
+#############################################################################
+event zeek_init() &priority=5 {
+  Analyzer::register_for_ports(Analyzer::ANALYZER_LDAP_TCP, LDAP::ports_tcp);
+  Analyzer::register_for_ports(Analyzer::ANALYZER_LDAP_UDP, LDAP::ports_udp);
+
+  Log::create_stream(LDAP::LDAP_LOG, [$columns=Message, $ev=log_ldap, $path="ldap", $policy=log_policy]);
+  Log::create_stream(LDAP::LDAP_SEARCH_LOG, [$columns=Search, $ev=log_ldap_search, $path="ldap_search", $policy=log_policy_search]);
+}
+
+#############################################################################
+function set_session(c: connection, message_id: int, opcode: LDAP::ProtocolOpcode) {
+
+  if (! c?$ldap_messages )
+    c$ldap_messages = table();
+
+  if (! c?$ldap_searches )
+    c$ldap_searches = table();
+
+  if ((opcode in OPCODES_SEARCH) && (message_id !in c$ldap_searches)) {
+    c$ldap_searches[message_id] = [$ts=network_time(),
+                                   $uid=c$uid,
+                                   $id=c$id,
+                                   $message_id=message_id,
+                                   $result_count=0];
+
+  } else if ((opcode !in OPCODES_SEARCH) && (message_id !in c$ldap_messages)) {
+    c$ldap_messages[message_id] = [$ts=network_time(),
+                                   $uid=c$uid,
+                                   $id=c$id,
+                                   $message_id=message_id];
+  }
+
+}
+
+#############################################################################
+@if (Version::at_least("5.2.0"))
+event analyzer_confirmation_info(atype: AllAnalyzers::Tag, info: AnalyzerConfirmationInfo) {
+  if ( atype == Analyzer::ANALYZER_LDAP_TCP ) {
+    info$c$ldap_proto = "tcp";
+  }
+}
+@else @if (Version::at_least("4.2.0"))
+event analyzer_confirmation(c: connection, atype: AllAnalyzers::Tag, aid: count) {
+@else
+event protocol_confirmation(c: connection, atype: Analyzer::Tag, aid: count) {
+@endif
+
+  if ( atype == Analyzer::ANALYZER_LDAP_TCP ) {
+    c$ldap_proto = "tcp";
+  }
+
+}
+@endif
+#############################################################################
+event LDAP::message(c: connection,
+                    message_id: int,
+                    opcode: LDAP::ProtocolOpcode,
+                    result: LDAP::ResultCode,
+                    matched_dn: string,
+                    diagnostic_message: string,
+                    object: string,
+                    argument: string) {
+
+  if (opcode == LDAP::ProtocolOpcode_SEARCH_RESULT_DONE) {
+    set_session(c, message_id, opcode);
+
+    if ( result != LDAP::ResultCode_Undef ) {
```

Comment:
This might look a bit more tidy if `ldap_searches[message_id]` and `ldap_messages[message_id]` was stored in a local variable upfront. Perf is another argument, through without a large test pcap hard to quantify.

Nothing to address, seems a consistent pattern across all event handlers.